### PR TITLE
Change API and reorganize DBSCAN structure

### PR DESCRIPTION
**Pull request by aprokop**
Trying to align DBSCAN API with [Python DBSCAN package](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.DBSCAN.html) and [R DBSCAN package](https://cran.r-project.org/web/packages/dbscan/index.html). The goal is to simplify the main `dbscan()` call API, and allow to easily extend its functionality.

- Main `dbscan` call signature was aligned more with other implementations
- A new `DBSCAN::Parameters` was introduced to expand functionality later
- Main `dbscan` call now returns labels
  This is similar to dbscan implementations in Python and R
- Verify was moved out of the main dbscan call
- Computing cluster CRS graph was now moved into the main file
  It is unclear whether that's the final form, but as it is unknown, it
  should not live in ArborX space right now
- Noise is marked as `-1`
  This introduced an additional kernel launch and extra memory array (`cluster_sizes`). I have not observed any noticable difference in performance.

Remaining things to discuss:
- Both Python and R packages return a DBSCAN structure as a result to `dbscan` call. The structure contains `eps`, `minpts`, `labels`, and possibly `core_indices`. I'm not sure why. It may be because of their ML nature, somehow.
- Should `labels` be passed as a reference argument, or returned as view?
- Where should conversion of CRS graph live? I would be fine leaving it outside of `ArborX` namespace.
- Both Python and R packages allow `X` argument (primitives for us so far) to also be a distance matrix. I don't think it is necessary for us at this point.

### CODE REVIEW

**Comment by aprokop** [comment]
@dalg24 I would appreciate a quick review on this, as the next steps would be:
- implement `DBSCAN*` algorithm
- compare results of `DBSCAN*` with R package
- unit test similarly to sklearn
  Since API are very similar, it should be possible to reuse some (all?) of the tests

**Review by dalg24 — COMMENTED** [review]
What do you anticipate would go into `DBSCAN::Parameters`?  Some user-defined distance function?  What else?
(Side note timers do not belong there.  You can achieve what you want with Kokkos profiling tools.  We can deal with this later though)

Should we return the labels instead of taking them as parameter?  How do we expect users will use them?

**Comment by aprokop** [comment]
> What do you anticipate would go into `DBSCAN::Parameters`? Some user-defined distance function? What else?

- Algorithm (DBSCAN, DBSCAN*, or border points)
- Underlying index structure (BVH/kdtree/bruteforce)
- (maybe) metric function
- verbosity

>  (Side note timers do not belong there. You can achieve what you want with Kokkos profiling tools. We can deal with this later though)

While somewhat true, it is a matter of convenience. Eventually, once we have an algorithm we are fully satisfied with, the timers would be removed. But before that, they are very valuable.

> Should we return the labels instead of taking them as parameter? How do we expect users will use them?

That is one of the questions I have. Some packages (R and cuML) return both `labels` and `core_sample_indices` as a result of `DBSCAN.fit(X)` call. I don't know why they do that.

**Comment by aprokop** [comment]
@dalg24 I switched to returning labels. I don't think there's anything else to be done in this PR. The discussion about the timers is valuable, but I'm postponing making a decision on that.

**Comment by aprokop** [comment]
retest this please

File: examples/dbscan/ArborX_DBSCAN.hpp [hunk 1]
```diff
@@ -195,130 +193,49 @@ void dbscan(ExecutionSpace exec_space, Primitives const &primitives,
...
-  Kokkos::parallel_for("ArborX::DBSCAN::flatten_stat",
+  Kokkos::View<int *, MemorySpace> cluster_sizes(
+      "ArborX::dbscan::cluster_sizes", n);
+  Kokkos::parallel_for("ArborX::dbscan::finalize_labels",
                        Kokkos::RangePolicy<ExecutionSpace>(exec_space, 0, n),
                        KOKKOS_LAMBDA(int const i) {
                          // ##### ECL license (see LICENSE.ECL) #####
                          int next;
-                         int vstat = stat(i);
+                         int vstat = labels(i);
```
Comment (dalg24) [inline examples/dbscan/ArborX_DBSCAN.hpp:203]: you can also alias the view.  I seem to remember you insisted on the name when you originally wrote this.

**Review by dalg24 — COMMENTED** [review]
Are they situations where you run dbscan multiple times on the same input data?

**Comment by aprokop** [comment]
> Are they situations where you run dbscan multiple times on the same input data?

Right now there isn't. In principle, one could be interested in doing sweeps over the input data, varying `eps` or `minpts`.

File: examples/dbscan/ArborX_DBSCAN.hpp (hunk 1 above)
Comment (aprokop) [reply examples/dbscan/ArborX_DBSCAN.hpp:203]: Yeah, I could. But I decided to change, as I always had trouble remembering what `stat` was. And now with `labels` around, it would have become even more confusing. Does not really change the ECL part of the code, I'm ok with this kind of change here.

**Review by dalg24 — APPROVED** [review]

**Comment by aprokop** [comment]
@dalg24 Thanks for the review!
